spock-core/src/main/java/org/spockframework/runtime/ExpressionInfoValueRenderer.java

```diff
@@ -87,6 +89,9 @@ private String doRenderValue(ExpressionInfo expr) {
     String result = renderAsFailedStringComparison(expr);
     if (result != null) return result;
 
+    result = renderAsFailedSetEqualityComparison(expr);
```

Comment:
I would register a set of custom value renderer with 2 methods: `boolean canHandle()`  with the check and `String render()`. The `doRenderValue()` must not change every time.